src/chunk.c

```diff
@@ -1790,53 +1977,195 @@ ts_chunk_get_by_relid(Oid relid, int16 num_constraints, bool fail_if_not_found)
 
 	schema = get_namespace_name(get_rel_namespace(relid));
 	table = get_rel_name(relid);
-	return chunk_get_by_name(schema, table, num_constraints, fail_if_not_found);
+	return chunk_get_by_name(schema, table, fail_if_not_found);
 }
 
-/* Lookup a chunk_id from a chunk's relid.
- * Optimize with memoization
+Chunk *
+ts_chunk_get_by_id(int32 id, bool fail_if_not_found)
+{
+	ScanKeyData scankey[1];
+
+	/*
+	 * Perform an index scan on chunk id.
+	 */
+	ScanKeyInit(&scankey[0], Anum_chunk_idx_id, BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(id));
+
+	return chunk_scan_find(CHUNK_ID_INDEX, scankey, 1, CurrentMemoryContext, fail_if_not_found);
+}
+
+/*
+ * Simple scans provide lightweight ways to access chunk information without the
+ * overhead of getting a full chunk (i.e., no extra metadata, like constraints,
+ * are joined in). This function forms the basis of a number of lookup functions
+ * that, e.g., translates a chunk relid to a chunk_id, or vice versa.
+ */
+static bool
+chunk_simple_scan(ScanIterator *iterator, FormData_chunk *form, bool missing_ok)
+{
+	int count = 0;
+
+	ts_scanner_foreach(iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(iterator);
+		chunk_formdata_fill(form, ti->tuple, ti->desc);
+
+		if (!form->dropped)
+			count++;
+	}
+
+	Assert(count == 0 || count == 1);
+
+	if (count == 0 && !missing_ok)
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT), errmsg("chunk not found")));
+
+	return count == 1;
+}
+
+static bool
+chunk_simple_scan_by_name(const char *schema, const char *table, FormData_chunk *form,
+						  bool missing_ok)
+{
+	ScanIterator iterator;
+
+	if (schema == NULL || table == NULL)
+		return false;
+
+	iterator = ts_scan_iterator_create(CHUNK, AccessShareLock, CurrentMemoryContext);
+	init_scan_by_name(&iterator, schema, table);
+
+	return chunk_simple_scan(&iterator, form, missing_ok);
+}
+
+static bool
+chunk_simple_scan_by_relid(Oid relid, FormData_chunk *form, bool missing_ok)
+{
+	bool found = false;
+
+	if (OidIsValid(relid))
+	{
+		const char *table = get_rel_name(relid);
+
+		if (table != NULL)
+		{
+			Oid nspid = get_rel_namespace(relid);
+			const char *schema = get_namespace_name(nspid);
+
+			found = chunk_simple_scan_by_name(schema, table, form, missing_ok);
+		}
+	}
+
+	if (!found && !missing_ok)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("chunk with relid %u not found", relid)));
```

Comment:
What's about when `relid` is not valid? In such case the error can be improved.

Reply:
This is a rare case and it would actually say that relid is 0. IMO, it is not worth adding a custom error for this.

Reply:
You add the error, so I don't understand why it should not print a meaningful message. I think it is fine to keep the same error code.

Reply:
I guess I misunderstood what you are asking. Can you explain why the current error is not meaningful and how you'd like it to be changed?

Reply:
In my last comment I was thinking about the other case, where you applied the changes already. Sorry it is my fault.

Actually your first reply answers my question.